src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,345 @@
+        if metadata is None:
+            metadata = dict()
+        elif not isinstance(metadata, dict):
+            raise TypeError('*metadata*: expected dict or None, got %s' % type(metadata))
+
+        body = {
+            'name': key,
+            'mimeType': Backend.MIME_TYPE_BINARY,
+            'parents': [ self.folder['id'] ],
+            'properties': self._encode_metadata(metadata),
+        }
+        log.debug("metadata: {0}, body: {1}".format(metadata, body))
+        f = self._create_file(body)
+        self._delete_other(f)
+        return ObjectW(self.service, f)
+
+    @copy_ancestor_docstring
+    def clear(self):
+        log.debug("clear")
+        for f in self._list_files(self.folder, "id"):
+            self._delete_file(f)
+
+    @copy_ancestor_docstring
+    def contains(self, key):
+        log.debug("contains {0}".format(key))
+        f = self._lookup_file(key, self.folder)
+        return f is not None
+
+    @copy_ancestor_docstring
+    def delete(self, key, force=False):
+        log.debug("delete {0}".format(key))
+        f = self._lookup_file(key, self.folder)
+        if f is not None:
+            self._delete_file(f)
+        elif not force:
+            raise NoSuchObject(key)
+
+    @copy_ancestor_docstring
+    def list(self, prefix=''):
+        log.debug("list {0}".format(prefix))
+        for f in self._list_files(self.folder, "id, name"):
+            name = f['name']
+            if not prefix or name.startswith(prefix):
+                yield name
+
+    @copy_ancestor_docstring
+    def update_meta(self, key, metadata):
+        log.debug("update_meta {0}: {1}".format(key, metadata))
+        if not isinstance(metadata, dict):
+            raise TypeError('*metadata*: expected dict, got %s' % type(metadata))
+        self.copy(key, key, metadata)
+
+    @copy_ancestor_docstring
+    def copy(self, src, dest, metadata=None):
+        log.debug("copy {0} -> {1}: {2}".format(src, dest, metadata))
+        if not (metadata is None or isinstance(metadata, dict)):
+            raise TypeError('*metadata*: expected dict or None, got %s' % type(metadata))
+
+        f = self._lookup_file(src, self.folder)
+        if f is None:
+            raise NoSuchObject(src)
+        body = {
+            'name': dest,
+            'mimeType': Backend.MIME_TYPE_BINARY,
+            'parents': [ self.folder['id'] ],
+            'properties': self._encode_metadata(metadata),
+        }
+        log.debug("metadata: {0}, body: {1}".format(metadata, body))
+        new_f = self._copy_file(f['id'], body)
+        if src == dest:
+            self._delete_other(new_f)
+
+class ObjectR(io.BytesIO):
+    '''A Google Drive object opened for reading'''
+
+    def __init__(self, service, f, metadata):
+        super().__init__()
+        self.metadata = metadata
+        request = service.files().get_media(fileId=f['id'])
+        if int(f['size']) > 0:
+            d = apiclient.http.MediaIoBaseDownload(self, request)
```

Comment:
Yes, ObjectR/ObjectW currently keep the data in-memory - I noted this in the submission. I will fix it a bit later and it does not have big impact now because data chunks are not bigger than 10MB in size.